### ISSUE TITLE
Poke the library for issues

### ISSUE THREAD

**Issue by arybczak** [haskell-effectful/effectful#1]
The library seems to be more or less feature complete for the initial release, all that remains is writing documentation and implementing a few minor things such as making `cloneEnv` work for forked environments).

I'd like to kindly request @isovector, @KingOfTheHomeless and @lexi-lambda to have a look (if you find a bit of time) and tell me if there's something obvious I'm missing. You have tons of experience with various effect libraries, so I'd really appreciate your insight.

I've scoured issue trackers of `polysemy` and `eff` in search for issues that might arise with `effectful`, but examples of things that are confusing / impossible to write with `polysemy` generally "just work" with `effectful`, e.g. [this](https://github.com/polysemy-research/polysemy/issues/373) one (compare the simplicity of the `runBroker` implementation):

```haskell
import Control.Concurrent
import Control.Monad.IO.Class
import Data.Functor

import Effectful.Interpreter
import Effectful.Monad

nestedImpl :: Int -> (String -> (Bool -> IO ()) -> IO ()) -> IO ()
nestedImpl i cb = void . forkIO $ do
  threadDelay $ 1 * 1000 * 1000
  cb (show i) $ \b -> putStrLn $ "result " ++ show b

data Broker :: Effect where
  Subscribe :: Int -> (String -> m Bool) -> Broker m ()

subscribe :: Broker :> es => Int -> (String -> Eff es Bool) -> Eff es ()
subscribe channel = send . Subscribe channel

runBroker :: IOE :> es => Eff (Broker : es) a -> Eff es a
runBroker = interpretIO $ \run -> \case
  Subscribe channel cb -> liftIO $ do
    putStrLn $ "Subscribe: " ++ show channel
    nestedImpl channel $ \s icb -> icb =<< run (cb s)

prog :: (IOE :> es, Broker :> es) => Eff es ()
prog = do
  subscribe 1 (\_ -> pure True)
  subscribe 2 (\_ -> pure False)
  liftIO . threadDelay $ 3 * 1000 * 1000

runProg :: IO ()
runProg = runEff . runIOE . runBroker $ prog
```

```
>>> runProg 
Subscribe: 1
Subscribe: 2
result False
result True
```

It also doesn't have [this](https://github.com/hasura/eff/issues/12) problem:

```haskell
import Effectful.Error
import Effectful.Interpreter
import Effectful.Monad
import Effectful.Reader

data SomeEff :: Effect where
  SomeAction :: SomeEff m String

someAction :: SomeEff :> es => Eff es String
someAction = send SomeAction

data Ex = Ex deriving (Eq, Show)
instance Exception Ex

good1 :: IO (Either ([String], Ex) String)
good1 = runEff . runError . interpret (\SomeAction -> throwError Ex) $ do
  someAction `catchError` \_ Ex -> return "caught"

good2 :: IO String
good2 = runEff $ runReader "unlocaled" . interpret (\SomeAction -> ask) $ do
  local (\_ -> "localed") someAction
```

```
>>> good1
Right "caught"
>>> good2
"localed"
```

As far as safety of the API goes, I'm aware of one caveat: defining your usual GADT effects and using `Effectful.Interpreter` for handling them is perfectly safe, but API for defining primitive/statically dispatched effects is unsafe in a sense that `unsafeCoerce` from `Effectful.Internal.Env` might leak, as if you use `putEffect` with one `i` and then `getEffect` with a different `i`, the compiler will not prevent you from doing that and things will explode. But that API will come with a big fat warning and should only be used when you absolutely need maximum performance and know what you're doing, so I'm willing to make that compromise.

**Comment by KingoftheHomeless** [comment]
I'll have a look-through. I already have a bunch of initial thoughts, but I haven't gotten deep enough yet to try and suss out if there are any real potential issues.

I see a bit of stuff I find questionable -- for example, the `MonadIO` instance of `Eff` is gated behind an `IOE` effect, but `MonadThrow`/`MonadCatch`/`MonadMask` instances aren't gated by any effect. `interpret` and `interpretM` could probably be merged, as the usability of `interpretM` could be greatly improved if it's possible to have the natural transformation provided to the handler instead be represented through an effect, like [Polysemy does with the WIP `interpretNew` combinator.](https://github.com/polysemy-research/polysemy/blob/dc9d5afebd2ec2c717a24b71799e674c5ac364ff/src/Polysemy/Internal/Combinators.hs#L463)

I like that this library is just ReaderT IO based (although with a very strange, interesting implementation!) I feel like sacrificing coroutines, continuations and nondeterminism isn't that big of a deal. Plus, you could probably do an `Async` effect without major issues. A potential change is to make it ReaderT ST based (and have the state token type be restricted to `RealWorld` if you want to do IO stuff) -- that way you'd be able to run `Eff`s purely for testing purposes.

**Comment by KingoftheHomeless** [comment]
Wait, your effects pull double duty. They either represent group of actions you may perform -- like conventional effects -- or mutable pieces of state which are modified, like with your implementation of Reader, Writer, State. In the case of Error and IOE, they're just dummy values. That's what you mean by static vs. dynamic dispatch, isn't it? That's... _very_ weird. What happens if you'd, for example, `send` a `Writer` effect and then interpret it? That breaks, doesn't it -- which is why the constructors of `Writer` aren't exposed?
This feels misguided to me. Effects should -- in principle -- represent possible actions, and I'd _only_ focus on that direction if at all possible, e.g I'd represent `Writer` as:
```haskell
data Writer w :: Effect where
  Tell :: w -> Writer w m ()
  Listen :: m a -> Writer w m (w, a)
  Pass :: m (w -> w, a) -> Writer w m a
```
Instead of a piece of mutable state to be manipulated via `stateEffect`/`putEffect`/`localEffect`/`unsafeStateEnv`/`unsafeModifyEnv`. Any necessary magic should be done during interpretation rather than when performing the actions.

This would allow for multiple different interpretations, which is half the point of effect systems. This doesn't really matter for `Writer`, since it's so elaborate that interpretation is extremely inflexible, but it would matter for a `Writer` variant with only the `Tell` action. It _definitely_ matters for `State`; you don't want your `State` effect to be magic.

Of course, I can't be certain how easily all of the effects of this library can be modified to be "true" effects, or what performance impact it would have; but I feel like having magic effects not only muddles things, but also limits the use of those effects.

**Comment by arybczak** [comment]
Thanks for looking into it :slightly_smiling_face: 

> MonadThrow/MonadCatch/MonadMask instances aren't gated by any effect

Yes, that's conscious. `MonadThrow` tells you nothing as you can call `Control.Exception.throw` everywhere anyway, while `MonadMask` is needed for exception-safe code even if you don't plan on catching and throwing them. `MonadCatch` is in principle debatable, but `MonadMask` requires `MonadCatch`, so it has to be there.

> Plus, you could probably do an Async effect without major issues.

You don't even have to do that. you can use `lifted-async` and it just works, there is a [test](https://github.com/arybczak/effectful/blob/e59a3bc7fb0c409ff17e77f89cf05570978ac285/tests/Main.hs#L92) that checks that. That's one of the goals, interoperability with existing libraries.

> A potential change is to make it ReaderT ST based (and have the state token type be restricted to RealWorld if you want to do IO stuff) -- that way you'd be able to run Effs purely for testing purposes.

I'm afraid the internal monad needs to be IO for having first-class support for exceptions :thinking: 

But I don't think that loses you much. If you provide pure interpretations of your effects, you can still test the code just fine. Testing frameworks have access to `IO`, so it doesn't really matter if you get `a` or `IO a` out of `runEff` :thinking: 

> Wait, your effects pull double duty. They either represent groups of actions you may perform -- like conventional effects -- or mutable pieces of state which are modified, like with your implementation of Reader, Writer, State. In the case of Error and IOE, they're just dummy values. That's what you mean by static vs. dynamic dispatch, isn't it?

Correct.

> What happens if you'd, for example, send a Writer effect and then interpret it? That breaks, doesn't it -- which is why the constructors of Writer aren't exposed?

Yes, that's a particular version of what I described above. Nice that you caught that :+1: It's a bit unfortunate, but as I said, it's a compromise I'm willing to make to allow people to get speed if they need it.

> Effects should -- in principle -- represent possible actions, and I'd only focus on that direction if at all possible

I don't like paying for things I don't need. There is no point paying for dynamic dispatch up front if you are not going to use it. Check out benchmarks, static version of state is very close to the reference implementation, no dynamic dispatch will ever come close.

There is no dynamically dispatched `Writer` because I didn't write it yet. There is a dynamically dispatched `State` though in `Effectful.State.Dynamic` that delegates to appropriate static versions.

There is no other sensible interpretation for `Reader` than the current one, so it's static. Same for `Error`, because it's `IO` inside, so you can only short-circuit via exceptions (and as a bonus you get a call stack).

**Comment by isovector** [comment]
I don't have enough effects paged in these days to be of help, sorry! But I wish you the best with this library!

**Comment by isovector** [comment]
That being said, I find the benchmarks suspicious --- they say polysemy is running faster than fused-effects. But, worse, is that it seems that nobody* [learned from my mistake](https://reasonablypolymorphic.com/blog/mea-culpa/) to make multi-module benchmarks.

* Not directed at you, I've also read two papers on effects systems since then that also fell into this trap.

**Comment by arybczak** [comment]
> I wish you the best with this library!

Thanks :)

> I find the benchmarks suspicious --- they say polysemy is running faster than fused-effects.

Not in all cases. `polysemy` is slower than `fused-effects` with a shallow stack (only necessary effects), but faster with a deep stack (5 readers, necessary effects, 5 readers), which simulates more realistic scenario in which you're using a portion of your stack. This makes sense to me, because `mtl` also gets a lot slower with a deep stack and `fused-effects` sort of "inherits" overhead of `mtl` by making their monad polymorphic + adds more overhead with their `Algebra`.

> But, worse, is that it seems that nobody* learned from my mistake to make multi-module benchmarks.

I know about this caveat :) However, it's really about preventing GHC from doing specialization that it wouldn't have done in a real world scenario. Putting functions in different modules is one way to achieve that, but putting NOINLINE pragmas on functions you don't want specialized (which is done in benchmarks) gives you the same outcome.

**Comment by isovector** [comment]
Ahh, that makes lots of sense. Sorry for my sensationalism! Very excited to see where this library goes!

**Comment by lexi-lambda** [comment]
This is an interesting design—when I first saw it, I assumed it was just the “obvious” implementation of an effect system in terms of `ReaderT` over `IO`, with all the usual expressiveness restrictions that apply. But upon closer inspection, there’s at least a little more going on (and the differences seem entirely orthogonal to the aforementioned expressiveness restrictions).

I haven’t taken the time to look through things in depth yet, but I can at least offer a few initial thoughts:

  * When I looked at the documentation on Hackage, I saw you supported a `Coroutine` effect, which surprised me, since that ought to be impossible given your approach. However, I noticed a little later that you have since removed it. The commit message didn’t elaborate upon why exactly you chose to remove it, but I assume it was because you realized it was, in fact, impossible (without being very broken, at least). :)

  * The “static dispatch” mode that your library supports is intriguing, and I think it’s a good idea. I see no reason to believe other effect systems (that aren’t based on monad transformers) can’t support it in exactly the same fashion you do, and it seems worth considering as a minor performance optimization.

  * Your `Error` effect is very weird in that it requires thrown values to implement `Exception`. This is the exact same mistake the old `ErrorT` transformer made in `Control.Monad.Trans.Error`, which is why we all use the less helpfully-named `ExceptT` instead.

    Please don’t repeat the same mistake! It is very obnoxious to be restricted in this fashion, as sometimes you just want simple short-circuiting despite not really raising “errors”. You can of course still implement `Error` in terms of RTS exceptions, just wrap them in a wrapper type that implements `Exception`, instead.

  * Your `Error` effect is also currently subtly broken. Here’s a failing test case using your example `FileSystem` effect:

    ```haskell
    test_throwFromInterpret :: Assertion
    test_throwFromInterpret = runIOE $ do
      result <- runError @FsError $ runFileSystemPure M.empty $
        runError @FsError $ readFile "foo.txt"
      assertEqual_ "caught by outer handler" True (isLeft result)
    ```

    The exception thrown by `readFile` should be caught by the outer `runError` handler, *not* the inner one, since the inner `runError` handler is not visible at the point of the `runFileSystemPure` call. However, your implementation of `throwError` does not currently indicate which handler it is throwing to, so the inner handler catches it anyway, breaking effect encapsulation.

  * You provide `MonadReader`, `MonadState`, and `MonadWriter` typeclasses without functional dependencies, which I understand are intended to ease combining monad transformers with an `Eff` base monad. This is, of course, entirely reasonable, but I encourage explicitly noting in the documentation that these classes should only be used with *concrete monad transformer stacks*, or else all the performance benefits of using `Eff` over `mtl`-style constraints are completely lost! Without such a disclaimer, you risk confusing people into thinking that writing code against a polymorphic monad is an encouraged idiom.

Aside from these issues, your library looks pretty good to me. The sacrifice of `NonDet`, `Coroutine`, and related effects is unfortunate, but honestly, I have started to wonder if such effects are really worth the hassle. Avoiding them means operations like `catch` and `listen` are really quite simple to implement in a way that behaves sensibly, and I think that’s pretty important. I think this approach might honestly be the best compromise for programmers we currently know how to achieve, so thank you for taking the time to put it together.

**Comment by KingoftheHomeless** [comment]
Re: Alexis' point on `Error` via `Exception`: [I've solved these problems in `in-other-words` by leveraging `Unique`s](https://hackage.haskell.org/package/in-other-words-0.2.0.0/docs/src/Control.Effect.Internal.Error.html#OpaqueExc). Feel free to adapt that implementation.

**Comment by arybczak** [comment]
Thanks for looking into it @lexi-lambda :slightly_smiling_face: 

> When I looked at the documentation on Hackage, I saw you supported a Coroutine effect, which surprised me, since that ought to be impossible given your approach. However, I noticed a little later that you have since removed it. The commit message didn’t elaborate upon why exactly you chose to remove it, but I assume it was because you realized it was, in fact, impossible (without being very broken, at least). :)

Indeed. I had a thought it could be made to work (for some definition of work) if the suspended computation could be resumed only once, but then I realized I have more important things to deal with  ;)

You're right about the `Exception` constraint in the `Error` effect, I already forgot about the `ErrorT` fiasco. I removed the `Exception` requirement and fixed the handler issue (thanks @KingoftheHomeless for the pointer and realization that the error can escape the scope of `runError` with `Async`).

> You provide MonadReader, MonadState, and MonadWriter typeclasses without functional dependencies, which I understand are intended to ease combining monad transformers with an Eff base monad.

My planned use case was for gradual transition from MTL style - first you replace your monadic stack with `Eff` (while your functions are still polymorphic in the monad), then you convert functions to use `Eff` instead and be polymorphic only in `es`. Though I'm pretty sure I'm gonna need to use GHC API to rewrite type signatures, doing this by hand with a large project is not gonna fly.

**Comment by mmhat** [comment]
Pinging @noughtmare and @xnning since I would be interested in their feedback as well.

**Comment by noughtmare** [comment]
Cool, I didn't expect to get pinged for this, because I am probably much less experienced than the other participants. I am personally more interested in the theoretical limits of the expressive power of effect systems, so this point in the design space is probably not for me, but I can imagine that it suffices for many practical applications. 

One thing that I might be able to add is that I think the "effects with non-linear control flow" (from the readme) corresponds with non tail-resumptive effects in the eveff papers. So, this library only seems to support tail-resumptive effects. Maybe that could be mentioned somewhere. I think eveff was designed to be fast for tail-resumptive effects, but I think that still requires built-in delimited continuations that @lexi-lambda was working on.

Also from the readme:

> Arguably it's a small price to pay for predictability, because such specialized effects are needed rarely and locally, at which point a dedicated, well established solution such as conduit, list-t or logict can be used.

I think the scoped effects paper gives some convincing examples like adding a counter to a nondeterminism effect to count the number of choices that were made during the computation. To me that sounds like it can be very useful for instrumenting effectful programs for the purpose of debugging and profiling.

Nevertheless, this library does looks very practical, good luck!

**Comment by arybczak** [comment]
Thanks everyone for the input :slightly_smiling_face:  If there's anything you'd like to bring in the future, please open a separate ticket.
